sqlobject/src/main/java/org/jdbi/v3/sqlobject/customizers/BindList.java

```diff
@@ -0,0 +1,178 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.sqlobject.customizers;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.lang.reflect.Array;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+import org.jdbi.v3.sqlobject.SqlStatementCustomizer;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizerFactory;
+import org.jdbi.v3.sqlobject.SqlStatementCustomizingAnnotation;
+import org.jdbi.v3.sqlobject.internal.ParameterUtil;
+
+/**
+ * Binds each value in the annotated {@link Iterable} or array/varargs argument, and defines a named attribute as a
+ * comma-separated list of each bound parameter name. Common use cases:
+ * <p>
+ * <pre>
+ * &#64;SqlQuery("select * from things where id in (&lt;ids&gt;)")
+ * List&lt;Thing&gt; getThings(@BindList int... ids)
+ *
+ * &#64;SqlQuery("insert into things (&lt;columnNames&gt;) values (&lt;values&gt;)")
+ * void insertThings(@DefineList List&lt;String&gt; columnNames, @BindList List&lt;Object&gt; values)
+ * </pre>
+ * <p>
+ * <p>
+ * Throws IllegalArgumentException if the argument is not an array or Iterable. How null and empty collections are handled can be configured with onEmpty:EmptyHandling - throws IllegalArgumentException by default.
+ * </p>
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER})
+@SqlStatementCustomizingAnnotation(BindList.Factory.class)
+public @interface BindList {
+    /**
+     * The attribute name to define. If omitted, the name of the annotated parameter is used. It is an error to omit
+     * the name when there is no parameter naming information in your class files.
+     *
+     * @return the attribute name.
+     */
+    String value() default "";
+
+    /**
+     * what to do when the argument is null or empty
+     */
+    EmptyHandling onEmpty() default EmptyHandling.THROW;
+
+    final class Factory implements SqlStatementCustomizerFactory {
+        @Override
+        public SqlStatementCustomizer createForParameter(Annotation annotation, Class<?> sqlObjectType, Method method, Parameter param, Object arg) {
+            final BindList bindList = (BindList) annotation;
+            final String name = ParameterUtil.getParameterName(bindList, bindList.value(), param);
+
+            if (arg == null || Util.size(arg) == 0) {
```

Comment:
Kinda crappy to potentially walk the entire list just to see if it's empty or not here.